src/test/java/com/shapesecurity/shift/es2016/Test262/PassTest.java

```diff
@@ -0,0 +1,439 @@
+			assertEquals(expectedComment.end.offset, actualComment.end.offset);
+		}
+
+		// check early error checker
+		ImmutableList<EarlyError> earlyErrors = EarlyErrorChecker.validate(actual);
+		if (earlyErrors.isNotEmpty()) {
+			throw new RuntimeException("Pass test throws early error: " + earlyErrors.maybeHead().fromJust().message);
+		}
+
+		// check validator
+		ImmutableList<ValidationError> validationErrors = Validator.validate(actual);
+		if (validationErrors.isNotEmpty()) {
+			throw new RuntimeException("Pass test fails to validate: " + validationErrors.maybeHead().fromJust().message);
+		}
+
+		// check codegen
+		String codegened = CodeGen.codeGen(actual);
+		Program codegenedAndParsed = name.endsWith(".module.js") ? parser.parseModule(codegened) : parser.parseScript(codegened);
+		assertTreesEqual(actual, codegenedAndParsed);
+
+		// check serializer and deserializer
+		String serialized = Serializer.serialize(actual);
+		Program deserialized = (Program) Deserializer.deserialize(serialized);
+		assertTreesEqual(actual, deserialized);
+	}
+
+	@Parameterized.Parameters(name = "{0}")
+	public static Collection<Object[]> data() {
+		File[] files = (new File(testsDir)).listFiles();
+		return Arrays.stream(files)
+				.map(f -> new Object[]{ f.getName() })
+				.collect(Collectors.toList());
+	}
+
+	@Parameterized.Parameter
+	public String name;
+
+	@Test
+	public void test() throws Exception {
+		XFailHelper.wrap(this.name, xfail, PassTest::check);
+	}
+
+	static class DeserializerWithLocation extends Deserializer {
+		protected HashTable<Node, SourceSpan> locations = HashTable.emptyUsingIdentity();
+
+		@Nonnull
+		public Maybe<SourceSpan> getLocation(@Nonnull Node node) {
+			return this.locations.get(node);
+		}
+
+		public DeserializerWithLocation() {}
+
+		@Override
+		protected Node deserializeNode(JsonElement jsonElement) throws ClassNotFoundException, IllegalAccessException, InstantiationException, NoSuchMethodException {
+			Node result = super.deserializeNode(jsonElement);
+			if (result != null) {
+				JsonObject jsonObject = jsonElement.getAsJsonObject();
+				if (jsonObject.has("loc")) {
+					JsonObject loc = jsonObject.getAsJsonObject("loc");
+					JsonObject start = loc.getAsJsonObject("start");
+					SourceLocation startLoc = new SourceLocation(
+							start.getAsJsonPrimitive("line").getAsInt(),
+							start.getAsJsonPrimitive("column").getAsInt(),
+							start.getAsJsonPrimitive("offset").getAsInt()
+					);
+
+					JsonObject end = loc.getAsJsonObject("end");
+					SourceLocation endLoc = new SourceLocation(
+							end.getAsJsonPrimitive("line").getAsInt(),
+							end.getAsJsonPrimitive("column").getAsInt(),
+							end.getAsJsonPrimitive("offset").getAsInt()
+					);
+
+					locations = locations.put(result, new SourceSpan(Maybe.empty(), startLoc, endLoc));
+				}
+			}
+			return result;
+		}
+	}
+
+	public static ImmutableList<ParserWithLocation.Comment> deserializeComments(String toDeserialize) {
```

Comment:
Since this parameter is just a `String` type, maybe call it something that hints at the format of the data that it contains.

Reply:
Actually, we could just take a `JsonElement` like `deserializeNode`.